File: python/whylogs_v1/core/column_profile.py
import logging
from typing import Any, Dict, List

from .preprocessing import PreprocessColumn
from .projectors import SingleFieldProjector
from .proto import ColumnMessage
from .schema import ColumnSchema
from .view import ColumnProfileView

logger = logging.getLogger(__name__)


class ColumnProfile(object):
    def __init__(self, name: str, schema: ColumnSchema, cache_size: int):
        self._name = name
        self._schema = schema

        self._trackers = schema.get_trackers(name)
        self._projector = SingleFieldProjector(col_name=name)
        self._success_count = 0
        self._failure_count = 0
        logger.debug("Setting cache size for column: %s with size: %s", self._name, cache_size)
        self._cache_size = cache_size
        self._cache: List[Any] = []

    def track(self, row: Dict[str, Any]) -> None:
        value = self._projector.apply(row)
        if len(self._cache) < self._cache_size - 1:
            self._cache.append(value)
        else:
            self._cache.append(value)
            cache_list = self._cache
            self._cache = []
            self.track_column(cache_list)

    def flush(self) -> None:
        """Force emptying the cache and update the internal metrics."""
        old_cache = self._cache
        self._cache = []
        self.track_column(old_cache)

    def track_column(self, series: Any) -> None:
        ex_col = PreprocessColumn.apply(series)
        for tracker in self._trackers.values():
            res = tracker.track_column(ex_col)
            self._success_count += res.failures
            self._failure_count += res.successes

    def serialize(self) -> ColumnMessage:
        self.flush()
        res = {}
        for t_name, t in self._trackers.items():
            res[t_name] = t.serialize()
        return ColumnMessage(trackers=res)

    def view(self) -> ColumnProfileView:
        metrics = {}
        for t_name, tracker in self._trackers.items():
            metrics[t_name] = tracker.to_mergeable()
        return ColumnProfileView(
            metrics=metrics,
            success_count=self._success_count,
            failure_count=self._failure_count,
        )

File: java/core/src/main/java/com/whylogs/core/ColumnProfile.java
package com.whylogs.core;

import com.whylogs.core.message.ColumnMessage;
import com.whylogs.core.metrics.Metric;
import com.whylogs.core.metrics.OperationResult;
import com.whylogs.core.schemas.ColumnSchema;
import com.whylogs.core.views.ColumnProfileView;
import java.util.ArrayList;
import java.util.HashMap;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@ToString
@EqualsAndHashCode
public class ColumnProfile<T> implements AutoCloseable {
  // Question: Perhaps it would be good to use a builder pattern here? (Instead of the init method)

  // Required
  private final String name;
  private final ColumnSchema schema;
  private final int cacheSize;

  // Has Defaults
  private HashMap<String, Metric<?>> metrics;
  private final SingleFieldProjector<T> projector;
  private int successCount;
  private int failureCount;
  private int nullCount;

  private ArrayList<T> cachedValues;

  public ColumnProfile(String name, ColumnSchema schema, int cacheSize) {
    this.name = name;
    this.schema = schema;
    this.cacheSize = cacheSize; // TODO: add logger for size of cache on column

    // Defaulted
    this.metrics = this.schema.getMetrics();
    this.projector = new SingleFieldProjector<>(name);
    this.successCount = 0;
    this.failureCount = 0;
    this.nullCount = 0;
    this.cachedValues = new ArrayList<>();
  }

  public void addMetric(Metric<?> metric) {
    if (this.metrics.containsKey(metric.getNamespace())) {
      // TODO: Add logger with warning about replacement
    }

    this.metrics.put(metric.getNamespace(), metric);
    // TODO: Wouldn't this implement a success count here?
  }

  // TODO: this only gets one not every part of the row. Should projector actually do it multiple?
  public void track(HashMap<String, Object> row) {
    T value = this.projector.apply(row);
    this.cachedValues.add(value);

    if (this.cachedValues.size() >= this.cacheSize) {
      this.flush();
    }
  }

  public void flush() {
    // TODO: Logger was initially here, but only for when it was forced, think it through
    ArrayList<T> oldCache = this.cachedValues;
    this.cachedValues = new ArrayList<>();
    this.trackColumn(oldCache);
  }

  public void trackColumn(ArrayList<?> values) {
    PreprocessedColumn proccessedColumn = PreprocessedColumn.apply(values);

    for (Metric<?> metric : this.metrics.values()) {
      OperationResult result = metric.columnarUpdate(proccessedColumn);
      this.successCount += result.getSuccesses();
      this.failureCount += result.getFailures();
      this.nullCount += result.getNulls();
    }
  }

  public ColumnProfileView view() {
    this.flush();
    return new ColumnProfileView(this.metrics, this.successCount, this.failureCount);
  }

  @Override
  public void close() throws Exception {
    this.flush();
  }
}
